invoices/invoiceregistry.go

```diff
@@ -676,25 +676,50 @@ func (i *InvoiceRegistry) processKeySend(ctx invoiceUpdateCtx) error {
 		return errors.New("final expiry too soon")
 	}
 
+	// The invoice database indexes all invoices by payment address, however
+	// legacy keysend payment do not have one. In order to avoid a new
+	// payment type on-disk wrt. to indexing, we'll continue to insert a
+	// blank payment address which is special cased in the insertion logic
+	// to not be indexed. In the future, once AMP is merged, this should be
+	// replaced by generating a random payment address on the behalf of the
+	// sender.
+	payAddr := channeldb.BlankPayAddr
+
 	// Create placeholder invoice.
 	invoice := &channeldb.Invoice{
 		CreationDate: i.cfg.Clock.Now(),
 		Terms: channeldb.ContractTerm{
 			FinalCltvDelta:  finalCltvDelta,
 			Value:           amt,
 			PaymentPreimage: &preimage,
+			PaymentAddr:     payAddr,
 			Features:        features,
 		},
 	}
 
 	// Insert invoice into database. Ignore duplicates, because this
 	// may be a replay.
 	_, err = i.AddInvoice(invoice, ctx.hash)
-	if err != nil && err != channeldb.ErrDuplicateInvoice {
+	switch {
+	case err == channeldb.ErrDuplicateInvoice:
+		dbInvoice, err := i.LookupInvoice(ctx.hash)
```

Comment:
I don't understand why this is necessary? For legacy keysends, no more insertions are happening in the payment addr index, so also no possibility of a collision?

Reply:
oops this wasn't supposed to be committed haha

Reply:
technically you can still get `ErrDuplicateInvoice` as a result of a replay or someone sending two keysends with the same payment hash. it is meant as a sanity check that this htlc is indeed the one that generated the invoice (primarily in the latter case)